File: ubuntu/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>



int main()
{

long number;
    // get card number
    do
    {
    number = get_long("Number: ");
    }
    while (number < 0);
    
    printf("%li\n", number);
    
    
    // determine card length
    
    long temp_number = number;
    int cardLength = 0;
    
    while (temp_number > 0)
    {
        temp_number = temp_number / 10;
        cardLength ++;
    }

// card length check
    printf("Card Length: %i \n", cardLength);   
    
    // check if card length is invalid
    if (cardLength != 15 && cardLength != 16 && cardLength != 13)
    {
        printf("INVALID\n");
    }
    else 
    {
    // turn card numbers into individual digits array
        int i;
        int lastDigit[cardLength];
    
        for (i = 0; i < cardLength; i++)
        {
            lastDigit[i] =  number % 10;
            number = number / 10;
        }
    
    // get *2 of every other digit and sum
        int k;
        int productDigitSum = 0;
        for (k = 1; k <= cardLength; k += 2)
        {
            if ((lastDigit[k] * 2) >= 10)
            {
                productDigitSum += ((lastDigit[k] * 2) % 10);
                productDigitSum += 1;
            }
            else 
            {
                productDigitSum += ((lastDigit[k] * 2) % 10);
            }
        }
        printf("ProductDigitsum: %i\n", productDigitSum);
        
        // get sum of non-summed digits above
        int l;
        int nonSummedNumbers = 0;
        for (l = 0; l <= cardLength; l+=2)
        {
            nonSummedNumbers += lastDigit[l];
        }
        
        // non summed digits check
        printf("Non Summed Digits: %i\n", nonSummedNumbers);
        
        int Total_Sums = productDigitSum + nonSummedNumbers;
        printf("Total of both loops: %i\n", Total_Sums);
          
        //check card type
        if ((Total_Sums % 10) == 0 && cardLength == 15)
        {
            if (lastDigit[14] == 3 && (lastDigit[13] == 4 || lastDigit[13] == 7))
            {
                printf("AMEX\n"); 
            }
            else
            {
                printf("INVALID\n");
            }
                
        } 
        else if ((Total_Sums % 10) == 0 && cardLength == 13)
        {
           printf("VISA\n"); 
        }
        else if ((Total_Sums % 10) == 0 && cardLength == 16)
        {
            if (lastDigit[15] == 5 && (lastDigit[14] == 1 || lastDigit[14] ==  2 || lastDigit[14] ==  3 || lastDigit[14] ==  4 || lastDigit[14] ==  5))
            {
                printf("MASTERCARD\n");
            }
            else if (lastDigit[15] == 4)
            {
                printf("VISA\n");
            }
            else 
            {
            printf("INVALID\n");
            }
        }
        else 
        {
            printf("INVALID\n");
        }
        
    }
    
}

File: ubuntu/pset6/credit/credit.py
from cs50 import get_int
import sys

def main():
    
    card = get_card()
    
    # check that we have card number
    #print(f"Card Number: {card}")
    
    # turn card number into a string for length and digit manipulation
    card_string = str(card)
    card_length = len(card_string)
    
    if (card_length != 15) and (card_length != 16) and (card_length != 13):
        print("INVALID")
        sys.exit(1)
       
    # declare product sum
    productSum = 0    
    
    #Find the sum of digits of every second number starting from 1 from the last
    for x in range((card_length - 2), -1, -2):
        product = int(card_string[x]) * 2
        if product >= 10:
            #print(f"{product}")
            last_digit= product % 10
            productSum += (1 + last_digit)
            
        else:
            #print(f"{product}")
            productSum += product
    
    #print(f"Second Number Digits {productSum}")
    
    #find the sum of non-previously summed digits
    for y in range((card_length - 1), -1, -2):
        product = int(card_string[y])
        productSum += product
        
    #print(f"Card passed productSum test with {productSum} which modulo 10s to: {productSum % 10}")
    
    if (productSum % 10) != 0:
        print("INVALID")
        sys.exit(0)
    
    else:
        if (int(card_string[0]) == 3 and int((card_string[1]) == 4 or int(card_string[1]) == 7)):
            print("AMEX")
        
        elif (int(card_string[0]) == 4):
            print("VISA")
            
        elif (int(card_string[0]) == 5 and (int(card_string[1]) == 1 or int(card_string[1]) == 2 or int(card_string[1]) == 3 or int(card_string[1]) == 4 or int(card_string[1]) == 5)):
            print("MASTERCARD")
            
        else:
            print("INVALID")
    
    
    
    
def get_card():
    while True:
        n = get_int("Number: ")
        if n > 0:
            break
    return n
    
    
main()